add multi arch docker build

b-reich opened the pull request:
Closes #4089

Changes proposed in this pull request:

- use buildx for docker building
- use github action for building

How to test the feature manually:

1.
2.
3.

Pull request checklist:

- [x] clear commit messages
- [ ] code manually tested
- [ ] unit tests written (optional if too hard)
- [ ] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/edge/docs/en/developers/04_Pull_requests.md).

b-reich committed:
Add multi arch container build

Co-authored-by: EdJoPaTo <rfc-conform-git-commit-email@funny-long-domain-label-everyone-hates-as-it-is-too-long.edjopato.de>
Co-authored-by: Alexandre Alapetite <alexandre@alapetite.fr>

b-reich commented:
Is publishing on github actions wantend?

b-reich commented:
https://github.com/b-reich/FreshRSS/actions/runs/6733089164/job/18301035008

looks like the build itself works

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,70 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
```
```suggestion
        uses: actions/checkout@v4
```

Alkarex commented:
Thanks, that looks pretty good already 👍🏻 
I am having a look

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,70 @@
+    branches:
+      - edge
+      - latest
+    tags:
+      - '[0-9]+.[0-9]+.[0-9]+'
+  pull_request:
+  workflow_dispatch:
```
I think we should have on `release` instead

b-reich replied:
I dont mind, was just an idea

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,70 @@
+  push:
+    branches:
+      - edge
+      - latest
```
Only for edge branch, not latest (which is done by `release` instead)

b-reich replied:
Done

Alkarex commented:
For reference, the current Docker Hub build configuration:

![image](https://github.com/FreshRSS/FreshRSS/assets/1008324/b5e78f73-6237-4771-8593-29640f68e6b8)

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,70 @@
+        id: docker_build_release
+        uses: docker/build-push-action@v5
+        with:
+          file: Docker/Dockerfile
```
It would be nice to also build `Docker/Dockerfile-Alpine`

b-reich replied:
Done

EdJoPaTo replied:
Just a random question, maybe there is a good answer here I am not seeing: Why multiple prebuilt images? Why not concentrate on one image and test that one instead of multiple ones? The image size difference is not that significant so personally I would concentrate on one image. (But thats personal taste, maybe you have different opinions and thats totally fine)

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,70 @@
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+  push-docker-readme:
```
This part is great, as I have not been able to push the correct Readme using solely Docker Hub...
We have a Docker readme at https://github.com/FreshRSS/FreshRSS/blob/edge/Docker/README.md but we could make a better page

b-reich replied:
Using the Docker Readme now

b-reich commented:
> For reference, the current Docker Hub build configuration:
> 
> ![image](https://user-images.githubusercontent.com/1008324/280092140-b5e78f73-6237-4771-8593-29640f68e6b8.png)

in the end we should only have something like latest and latest-alpine i think. No need to tag the arm images separately. This is kinda a breaking change than.

EdJoPaTo commented:
Maybe the Dockerfile Labels should also be adapted for this. The `docker/metadata-action` sets a bunch of labels by itself so we shouldn't duplicate ourselves here?

https://github.com/FreshRSS/FreshRSS/blob/8b2f7848eb1054ee26048e3ffe76855410286407/Docker/Dockerfile-Alpine#L23-L33

b-reich commented:
> Maybe the Dockerfile Labels should also be adapted for this. The `docker/metadata-action` sets a bunch of labels by itself so we shouldn't duplicate ourselves here?
> 
> https://github.com/FreshRSS/FreshRSS/blob/8b2f7848eb1054ee26048e3ffe76855410286407/Docker/Dockerfile-Alpine#L23-L33

The generated one from the metadata action are not used at the moment. this could be change @Alkarex what do you think?

Alkarex commented:
> No need to tag the arm images separately.

Indeed. I think it would be an acceptable breaking change.

> Why multiple prebuilt images?

Providing a Debian and an Alpine version is quite common. There are pros and cons with both (for instance build speed for Alpine, among other things), and having an alternative can be helpful (e.g. https://github.com/FreshRSS/FreshRSS/discussions/5742 )

I will come back for the other questions

b-reich commented:
https://github.com/b-reich/FreshRSS/actions/runs/6738089530/job/18316933656

the github action is nearly finish I think....
How to handle the tags is a thing now.

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,86 @@
+    branches:
+      - edge
+  release:
+  workflow_dispatch:
```
Is that used?

b-reich replied:
You mean the workflow_dispatch? 
I like the option to start the pipeline per hand.
Do you think it's not needed?

What do you prefer the on release or on tag? (The tag also happen on a release I think)

Alkarex replied:
I prefer to use `release`, as this is the way new official versions are produced, and yes, I also believe a release will also trigger a tag event.
Yes, I meant the `workflow_dispatch`. Yes, starting the pipeline by hand is always handy (haha), but I am not sure whether `workflow_dispatch` is needed for that.

b-reich replied:
Yeah its needed to define there
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch

Alkarex commented:
> The generated one from the metadata action are not used at the moment. this could be change @Alkarex what do you think?

Yes, it would probably be better to use the ones from the Action, indeed

b-reich committed:
using label from github action

Alkarex commented on Docker/Dockerfile:
```diff
@@ -23,18 +23,6 @@ ARG FRESHRSS_VERSION
 ARG SOURCE_BRANCH
 ARG SOURCE_COMMIT
 
-LABEL \
-	org.opencontainers.image.authors="Alkarex" \
-	org.opencontainers.image.description="A self-hosted RSS feed aggregator" \
-	org.opencontainers.image.documentation="https://freshrss.github.io/FreshRSS/" \
-	org.opencontainers.image.licenses="AGPL-3.0" \
-	org.opencontainers.image.revision="${SOURCE_BRANCH}.${SOURCE_COMMIT}" \
-	org.opencontainers.image.source="https://github.com/FreshRSS/FreshRSS" \
-	org.opencontainers.image.title="FreshRSS" \
-	org.opencontainers.image.url="https://freshrss.org/" \
-	org.opencontainers.image.vendor="FreshRSS" \
-	org.opencontainers.image.version="$FRESHRSS_VERSION"
```
I would prefer to keep the static ones here, so that they exist when the builds are done locally, e.g. like
https://github.com/FreshRSS/FreshRSS/blob/8b2f7848eb1054ee26048e3ffe76855410286407/Docker/freshrss/docker-compose.yml#L12-L15

`authors` can be removed

By the way, if anyone knows a way to populate `revision` and/or `version` even in those more manual scenarios, let me know

b-reich replied:
not sure what happen when you use the github action ones and the static ones.

Alkarex replied:
Let's try to remove the static ones from GitHub Actions, so that we can see whether that is sufficient or whether they need to be overridden again

b-reich replied:
I found some docu about. 
passing labels at build time will overwrite the static ones

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,91 @@
+      #     username: ${{ secrets.DOCKERHUB_USERNAME }}
+      #     password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      # - name: Login to GitHub Container Registry
```
Any comments about how pros and cons of (also) using the GitHub Container Registry?
Probably seems overkill for now?

b-reich replied:
For podman user?
And Dockerhub had in the pass some weird decisions I think.

Personaly I prefer the github registry

Alkarex replied:
Ok. Let's keep it but comment out for now, so we can easily activate if needed.
I am not familiar with the use-case: Can't Podman users pull from Docker Hub?

b-reich replied:
It can, but the default registry (dockerhub) is not set.

EdJoPaTo replied:
Newer podman versions try dockerhub on unspecified image tags. Personally I don't think it matters much as its the same image

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,91 @@
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v3
+
+      # - name: Login to DockerHub
+      #   uses: docker/login-action@v3
+      #   with:
+      #     username: ${{ secrets.DOCKERHUB_USERNAME }}
+      #     password: ${{ secrets.DOCKERHUB_TOKEN }}
```
I will take care or that

b-reich replied:
I enable it now.

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,91 @@
+    - uses: actions/checkout@v4
+
+    - name: Update repo description
+      uses: peter-evans/dockerhub-description@v3
```
For third-party Actions, I would prefer to pin to a specific git hash instead of `@v3`, to increase trust

b-reich replied:
I put it into a seperate action and pin it to a commit

b-reich committed:
dont remove static labels and split docker readme in seperate action

b-reich committed:
pin 3rd party action to a commit

b-reich committed:
enable pushing to dockerhub

EdJoPaTo commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,77 @@
+              suffix=-alpine
+              onlatest=true
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
```
The checkout can be removed from the action as long as `docker/build-push-action` has no context option defined. It will pull the repo by itself and has no need for a checkout.

EdJoPaTo commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,77 @@
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v3
+
+    - name: Login to DockerHub
+      uses: docker/login-action@v3
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+    # - name: Login to GitHub Container Registry
+    #   uses: docker/login-action@v3
+    #   with:
+    #     registry: ghcr.io
+    #     username: ${{ github.repository_owner }}
+    #     password: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Metadata
+      id: meta
+      uses: docker/metadata-action@v5
+      with:
+        flavor: ${{ matrix.flavor }}
+        images: |
+          docker.io/freshrss/freshrss
+      #     ghcr.io/${{ github.repository }}
+        tags: |
+          type=edge
+          type=semver,pattern={{version}}
+          type=semver,pattern={{major}}.{{minor}}
+          type=semver,pattern={{major}},enable=${{ !startsWith(github.ref, 'refs/tags/0.') }}
+        labels: |
+          org.opencontainers.image.description="A self-hosted RSS feed aggregator"
+          org.opencontainers.image.documentation="https://freshrss.github.io/FreshRSS/"
+          org.opencontainers.image.url="https://freshrss.org/"
```
Personally I do the login as a last step before the build for two reasons.
The login can be time critical, if I remember correctly it only lasts for an hour for the ghcr.

Also I order my actions based on what they generate for the following actions. The fewer information the following actions have the better. A login is something that could be used otherwise so expose as few actions afterwards with it as possible. Probably irrelevant most of the time but I still try to follow that when creating workflows.

```suggestion
    - name: Metadata
      id: meta
      uses: docker/metadata-action@v5
      with:
        flavor: ${{ matrix.flavor }}
        images: |
          docker.io/freshrss/freshrss
      #     ghcr.io/${{ github.repository }}
        tags: |
          type=edge
          type=semver,pattern={{version}}
          type=semver,pattern={{major}}.{{minor}}
          type=semver,pattern={{major}},enable=${{ !startsWith(github.ref, 'refs/tags/0.') }}
        labels: |
          org.opencontainers.image.description="A self-hosted RSS feed aggregator"
          org.opencontainers.image.documentation="https://freshrss.github.io/FreshRSS/"
          org.opencontainers.image.url="https://freshrss.org/"

    - name: Login to DockerHub
      uses: docker/login-action@v3
      with:
        username: ${{ secrets.DOCKERHUB_USERNAME }}
        password: ${{ secrets.DOCKERHUB_TOKEN }}

    # - name: Login to GitHub Container Registry
    #   uses: docker/login-action@v3
    #   with:
    #     registry: ghcr.io
    #     username: ${{ github.repository_owner }}
    #     password: ${{ secrets.GITHUB_TOKEN }}
```

EdJoPaTo commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,77 @@
+      - edge
+  release:
+  workflow_dispatch:
+
```
Restrict (or add) permissions to the job. When the ghcr is used the permission to write to the packages is needed. Always specify as less as possible here. The restricted token is also the default for new repos so it only works without in older repos. See [token permission documentation](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token)
```suggestion

permissions:
  contents: read
  packages: write

```

b-reich replied:
the token is not used add the moment. But a good addition

EdJoPaTo replied:
Maybe add the `packages: write` commented out like the ghcr login part is currently then?

EdJoPaTo commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,77 @@
+on:
+  push:
+    branches:
+      - edge
```
Personally I like to build on all branches and just not push the result to the registry then. It ensures the container still builds. But that's probably a decision for the maintainers.

Alkarex replied:
I am concerned by our energy footprint, so I try to limit the automated builds to a minimum.
When working on making changes to one of the Docker images, they need to be tested locally anyway, so auto-build does not add a lot of value. And if the builds fail once in a while, it is not a big deal as it does not affect production and can be fixed or reverted with another PR.

b-reich committed:
Update .github/workflows/build-images.yml

Co-authored-by: EdJoPaTo <github@edjopato.de>

b-reich committed:
remove not needed checkout

b-reich committed:
set github token permissions

Alkarex commented on .github/workflows/push-dockerhub-readme.yml:
```diff
@@ -0,0 +1,23 @@
+name: push readme to dockerhub
```
I think I would better like to have it back in the `build-images.yml`, mostly to be more resource efficient (avoid having to spin a new VM and git clone once more)

b-reich replied:
It was a seperated job
so it always create a new context on another machine

Alkarex replied:
Can't we just add the steps to the existing job?

Alkarex replied:
(It would be fine to react to the same events we already have)

Alkarex replied:
But good point, with your current suggestion, this job might not have to run that often.
Let's keep it that way then.

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,78 @@
+          type=semver,pattern={{major}}.{{minor}}
+          type=semver,pattern={{major}},enable=${{ !startsWith(github.ref, 'refs/tags/0.') }}
+        labels: |
+          org.opencontainers.image.description="A self-hosted RSS feed aggregator"
+          org.opencontainers.image.documentation="https://freshrss.github.io/FreshRSS/"
+          org.opencontainers.image.url="https://freshrss.org/"
```
So what was the conclusion? Can't those static labels be removed, as they are already defined in the Dockerfile?

b-reich replied:
This overwrite the metadata ones. And the metadata ones overwrite the statics ones.
So we benefit from the right versions label (and other labels)

Alkarex replied:
I am still confused: without these lines, what would `org.opencontainers.image.description` contain? Wouldn't it have what was defined in Dockefile?

b-reich replied:
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/28afa6b7-8825-43f1-bd89-5369850d0e0b)

b-reich replied:
when I dont set in manualy for the actions its taken from the github repo
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/0392d2d8-0872-4c79-90f1-b3b4b574ee85)

b-reich replied:
The metadata action set it to `A free, self-hostable news aggregator… `I overwrite it to `A self-hosted RSS feed aggregator`because the static ones in the dockerfile would be overwritten by the metadata ones.

Alkarex replied:
Ok, thanks for the investigation 👍🏻

Alkarex commented on .github/workflows/push-dockerhub-readme.yml:
```diff
@@ -0,0 +1,23 @@
+    paths:
+      - Docker/README.md
+    branches:
+      - latest
```
```suggestion
      - edge
```

Branch would be `edge` though

Alkarex commented:
Excellent job, this looks really nice 👍🏻 
We need a little update to https://github.com/FreshRSS/FreshRSS/blob/edge/Docker/README.md#our-docker-image-variants
Next step from my side will be to disable the `edge` builds on Docker Hub to see how well it behaves, and then later disable the other tags.

Alkarex committed:
Update .github/workflows/push-dockerhub-readme.yml

b-reich committed:
update docker readme to match new tags

b-reich commented:
@Alkarex please take a look on the new readme

Alkarex commented on Docker/README.md:
```diff
@@ -66,13 +66,16 @@ The [tags](https://hub.docker.com/r/freshrss/freshrss/tags) correspond to FreshR
+* `:x.y.z` tags correspond to [specific FreshRSS releases](https://github.com/FreshRSS/FreshRSS/releases), allowing you to target a precise version for deployment
+* `:x.y` tags are tied to a specific major version and minor version number. For example, `:12.2` will automatically receive updates for any `12.2.x` releases, but will not update to `12.3.x`
+* `:x` tags track the latest release within a major version series. For instance, `:12` will update to include any `12.x` releases, but will exclude versions beyond `13.x`
+* `*-alpine` use apline as base-image instead debian
```
```suggestion
* `*-alpine` use Linux Alpine as base-image instead of Debian
```

Alkarex commented on Docker/README.md:
```diff
@@ -66,13 +66,16 @@ The [tags](https://hub.docker.com/r/freshrss/freshrss/tags) correspond to FreshR
-* `:x.y.z` are [specific FreshRSS releases](https://github.com/FreshRSS/FreshRSS/releases)
-* `:arm` or `:*-arm` are the ARM `arm32v7` versions (e.g., for Raspberry Pi).
+* `:x.y.z` tags correspond to [specific FreshRSS releases](https://github.com/FreshRSS/FreshRSS/releases), allowing you to target a precise version for deployment
+* `:x.y` tags are tied to a specific major version and minor version number. For example, `:12.2` will automatically receive updates for any `12.2.x` releases, but will not update to `12.3.x`
+* `:x` tags track the latest release within a major version series. For instance, `:12` will update to include any `12.x` releases, but will exclude versions beyond `13.x`
```
```suggestion
* `:x.y` tags are tied to a specific major version and minor version number. For example, `:1.23` will automatically receive updates for any `1.23.x` releases, but will not update to `1.24.x`
* `:x` tags track the latest release within a major version series. For instance, `:1` will update to include any `1.x` releases, but will exclude versions beyond `2.x`
```

I have changed to examples closer to what we have in FreshRSS

Alkarex commented on Docker/README.md:
```diff
@@ -66,13 +66,16 @@ The [tags](https://hub.docker.com/r/freshrss/freshrss/tags) correspond to FreshR
 
 ### Linux: Debian vs. Alpine
 
-Our default image is based on [Debian](https://www.debian.org/). We offer an alternative based on [Alpine](https://alpinelinux.org/) (with the `:alpine` or `*-alpine` tag suffix).
```
Don't we have the `freshrss/freshrss:alpine` tag anymore?

b-reich replied:
when I am correct we have `latest-alpine`or `v2-alpine` and so on

b-reich replied:
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/da58321d-67b6-40f6-9ef4-1195830eb53e)
I testet it... there is something odd

b-reich replied:
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/20d2ed94-cb6c-4fdb-b3f9-96ba6792c004)
working now

Alkarex replied:
So `freshrss:latest-alpine`, but no `freshrss:alpine`?

b-reich replied:
yes

Alkarex replied:
Can you think of a way to have `freshrss:alpine` ? (and maybe no `freshrss:latest-alpine` at all)
It seems to be the more common way, e.g. https://hub.docker.com/layers/library/httpd/alpine/images/sha256-e4e78032a4c6375e4e697d68f9e70306bce66ccdf09659d4c06b029d5fc60d41?context=explore

b-reich replied:
done, but I find `latest-alpine` whould be "nice"

b-reich replied:
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/c42054ca-6746-4555-b6d8-49b745133695)

Alkarex committed:
Apply suggestions from code review

Alkarex committed:
Merge branch 'edge' into edge

Alkarex commented:
Ah, for the record, I was a bit confused because there were some actions triggered by pushes to this PR:
* https://github.com/b-reich/FreshRSS/actions/runs/6756095787
But it is because the name of @b-reich 's PR branch is also `edge`

https://github.com/orgs/community/discussions/26409#discussioncomment-3251822

b-reich committed:
fix suffix for alpine image

b-reich committed:
fix suffix for alpine image

Alkarex commented on .github/workflows/push-dockerhub-readme.yml:
```diff
@@ -0,0 +1,23 @@
+  workflow_dispatch:
+
+jobs:
+  push-docker-readme:
```
```suggestion
  push-docker-readme:
    if: github.repository_owner == 'FreshRSS'
```

https://github.com/orgs/community/discussions/26409#discussioncomment-3251822
I do not know whether that condition can be put higher up, in the `on:` section

b-reich replied:
already implemented with the other push logic

Alkarex commented on .github/workflows/build-images.yml:
```diff
@@ -0,0 +1,76 @@
+  # packages: write
+
+jobs:
+  build-container-image:
```
```suggestion
  build-container-image:
    if: github.repository_owner == 'FreshRSS'
```

https://github.com/orgs/community/discussions/26409#discussioncomment-3251822

Maybe that could be done smarter. It would be OK to let the build occur, as long as there is no push attempt

b-reich replied:
I add some logic to prevent the push on non upstream repos

b-reich committed:
push images only at upstream repo

b-reich committed:
push images only at upstream repo

b-reich committed:
push images only at upstream repo

b-reich committed:
tag latest-alpine as alpine

b-reich committed:
tag latest-alpine as alpine

b-reich committed:
remove no needed falvor

b-reich committed:
keep falvor for build

Alkarex committed:
Clean more things

Alkarex commented:
It looks almost good now :-)
I can spot one thing, which seems to be missing: we used to have the FreshRSS version in the `org.opencontainers.image.version` metadata:

https://github.com/FreshRSS/FreshRSS/blob/ecf1585d74c44954720acddce02d3827c93a06df/Docker/Dockerfile#L36

https://github.com/FreshRSS/FreshRSS/blob/ecf1585d74c44954720acddce02d3827c93a06df/Docker/hooks/build#L4

b-reich commented:
> It looks almost good now :-) I can spot one thing, which seems to be missing: we used to have the FreshRSS version in the `org.opencontainers.image.version` metadata:
> 
> https://github.com/FreshRSS/FreshRSS/blob/ecf1585d74c44954720acddce02d3827c93a06df/Docker/Dockerfile#L36
> 
> https://github.com/FreshRSS/FreshRSS/blob/ecf1585d74c44954720acddce02d3827c93a06df/Docker/hooks/build#L4

Do you think the default from the metadata action didnt feed the needs? It kinda match the labels now.
alpine:
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/63291375-cbc5-41fc-bbde-1254a2acd0b4)
debian:
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/88297de6-73c1-4e86-9701-d76170900e57)

Alkarex committed:
Rename action
To use same default name than https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml

Alkarex commented:
> Do you think the default from the metadata action didnt feed the needs? It kinda match the labels now.

It is very good for the releases, but we lack the FreshRSS version in the case of `edge`, don't we?

Alkarex commented:
Just for future inspiration (not for this PR), here is the default template for Docker Publish:
https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml
It is using e.g. https://github.com/sigstore/cosign , which I am not familiar with, but which sounds interesting

Alkarex committed:
Rename readme action
To use same default name than https://github.com/peter-evans/dockerhub-description/blob/main/.github/workflows/dockerhub-description.yml

b-reich commented:
@Alkarex its look like the cuurrent docker builds at the moment have a broken version label.
![grafik](https://github.com/FreshRSS/FreshRSS/assets/1573025/3c36af45-7df2-48f7-b971-1dfb8b690a35)

Yes the edge build have the version `edge` or `edge-alpine`. But its a rolling release. So its kinda personal taste I think.

Alkarex committed:
Use default name for dockerhub-description
https://github.com/peter-evans/dockerhub-description/blob/main/.github/workflows/dockerhub-description.yml
Plus minor wording

Alkarex commented:
> its look like the cuurrent docker builds at the moment have a broken version label.

Indeed 😱
Edit: That was a regression due to https://github.com/FreshRSS/FreshRSS/pull/5089

Alkarex committed:
Experiment with build-args

Alkarex committed:
Debug

b-reich commented:
@Alkarex  you need to checkout the repo when you try to retrive the version

Alkarex commented:
> you need to checkout the repo when you try to retrive the version

Thanks. Right, I had read a bit too fast that comment https://github.com/FreshRSS/FreshRSS/pull/5808#discussion_r1381382672

As you can see, I am experimenting with an arg-based approach for labels. What do you think?

Alkarex committed:
Add checkout back

Alkarex commented:
Hum, I was hopping that `docker/metadata-action` could be skipped, but `tags: ${{ matrix.tags }}` does not seem to work directly...

Alkarex committed:
Revert back to metadata-action

Alkarex committed:
Remove quotes in version

Alkarex committed:
Try to fix variables

Alkarex committed:
Experiment with automatic label values again

Alkarex commented:
The output looks good to me now, although I would have liked to simplify a bit more if we could avoid `docker/metadata-action`.
Final thoughts @b-reich ?

Alkarex committed:
Delete last Docker Hub hooks

EdJoPaTo commented:
Out of curiousity: why would you like to avoid using metadata-action? it seems helpful to me for generating the v1, v1.2, v1.2.3 tags and so on. Maybe I overlook something?

b-reich commented on .github/workflows/docker-publish.yml:
```diff
@@ -0,0 +1,94 @@
+        tags: ${{ matrix.tags }}
+        labels: |
+          org.opencontainers.image.revision=${{ github.ref_name }}.${{ github.sha }}
+          org.opencontainers.image.version=${{ env.FRESHRSS_VERSION }}
```
this is already handle by the `org.opencontainers.image.revision` label which is the commit hash
dont know if there is benefit to know which branch is build from, when you have the commit

also thinks the container version label is fine with sometthing like `1.22.2` or `1.22.2-alpine`its specify the container version not the version of the application itself

Alkarex replied:
> if there is benefit to know which branch is build from

This is mostly because this is what we already have, and I find it a bit easier to read with something like `1.22.1.cc2878aed855463efe36cf6fb96d001dbb75612e`. But I would not be much against leaving it to only the git SHA, if this is what is most standard.

Edit: indeed, according to https://specs.opencontainers.org/image-spec/annotations/#pre-defined-annotation-keys : **org.opencontainers.image.revision** Source control revision identifier for the packaged software.
I will edit accordingly

> also thinks the container version label is fine with sometthing like 1.22.2 or 1.22.2-alpineits specify the container version not the version of the application itself

I would also be happy with `1.22.2-alpine`, but that does not work for e.g. `edge` builds

Alkarex replied:
Switched back to default revision https://github.com/FreshRSS/FreshRSS/pull/5808/commits/3d110b8ec9af4ba9423890c1e018e99c8fbbaec1

b-reich commented:
Not sure why you would avoid using the metadata action?
Personally I wouldn't worry so much about the labels. Many people aren't even interested in them.

Alkarex commented:
> Out of curiousity: why would you like to avoid using metadata-action? it seems helpful to me for generating the v1, v1.2, v1.2.3 tags and so on. Maybe I overlook something?

Although it does not matter much, it would be mostly to remain as simple as possible, as we do not know how things will evolve in the future. For instance, we have been through periods during which the auto-build in Docker Hub was not available while waiting for a (initially long) *Sponsored OSS* process. During that period, I built and pushed images manually. The more tags there are, the more difficult it is to move to another build system, and the more work there is for the maintenance.

Another use-case: if we were to move back to building on Docker Hubs, there is poor support for multiple tags, which each take a new build, so I would not like to have 1, 1.22, 1.22.1 but only 1.22.1. Especially not when combined with multiple architectures and Debian + Alpine. It just becomes too many tags.

Furthermore, FreshRSS is not a library. And we do not have the resources to maintain several branches anyway. So the semver semantics only marginally applies to it. For instance, fixes in 1.22.1 may very well be in 1.23.0. So I am not sure why anyone would chose to track 1.22 instead of 1.22.1 or latest. And by offering semver tags, it might mislead some users into believing that we do have updates on old branches, which we do not.

> Personally I wouldn't worry so much about the labels. Many people aren't even interested in them.

Yes, and that is why I would probably like to err on the side of simplicity instead of having rarely-used features, which might end up having some drawbacks (see my explanation higher up). As soon as we introduce something, I take on me to actually make a good effort to maintain it over time, so I need to balance what we add with its actual value.

Alkarex committed:
Use only git SHA for org.opencontainers.image.revision
https://specs.opencontainers.org/image-spec/annotations/#pre-defined-annotation-keys

EdJoPaTo commented:
> Although it does not matter much, it would be mostly to remain as simple as possible, as we do not know how things will evolve in the future. For instance, we have been through periods during which the auto-build in Docker Hub was not available while waiting for a (initially long) _Sponsored OSS_ process. During that period, I built and pushed images manually. The more tags there are, the more difficult it is to move to another build system, and the more work there is for the maintenance.
> 
> Another use-case: if we were to move back to building on Docker Hubs, there is poor support for multiple tags, which each take a new build, so I would not like to have 1, 1.22, 1.22.1 but only 1.22.1. Especially not when combined with multiple architectures and Debian + Alpine. It just becomes too many tags.

When I moved away from auto build on Docker Hub it supported multiple tags like 1, 1.2, 1.2.3, … But that is some time ago. I explicitly searched for a way to replicate that behavior on GitHub Actions so I know and understand the pain points you are speaking of. Thanks for clarifying :)


> Furthermore, FreshRSS is not a library. And we do not have the resources to maintain several branches anyway. So the semver semantics only marginally applies to it. For instance, fixes in 1.22.1 may very well be in 1.23.0. So I am not sure why anyone would chose to track 1.22 instead of 1.22.1 or latest. And by offering semver tags, it might mislead some users into believing that we do have updates on old branches, which we do not.

Personally I use the main version tag for my containers. I get all the new features and bugfixed without worrying about breaking changes (new major release). Chosing a minor release would be to manually update but get bugfix only releases which don't change much and update the rest manually (like you suggested pinning a git hash on a used action.

Just to give an idea here: Maybe only support major and bugfix version (1 and 1.2.3). Minor releases and bugfix releases would be kinda the same. When someone wants to pin them they can. If they only care for breaking changes they can use the major version tag. This reduces the amount of tags while still allowing for non-breaking tags like I mainly use.
After all its only a single line less in the metadata-action config.

Alkarex committed:
Comment out semver for now
Might be enabled later if desired

Alkarex commented:
> Maybe only support major and bugfix version (1 and 1.2.3)

Ok, let's go for that

Alkarex committed:
Enable major semver

Alkarex commented:
Let's give it a spin 🚀
Thanks for your patience and constructive dialogue so far 👍🏻 

Seems to work fine for most of it:
https://github.com/FreshRSS/FreshRSS/actions/runs/6763929607/job/18381624080

![image](https://github.com/FreshRSS/FreshRSS/assets/1008324/d89e0236-7d37-4f71-a101-01580594e71f)

![image](https://github.com/FreshRSS/FreshRSS/assets/1008324/ade4876c-2026-4eed-a632-88a1e8f288f9)

There are only a few things that need to be ironed out:
https://github.com/FreshRSS/FreshRSS/actions/runs/6763929609/job/18381624094

Alkarex commented:
@b-reich and @EdJoPaTo : please add a line for you in https://github.com/FreshRSS/FreshRSS/blob/edge/CREDITS.md

Alkarex commented:
Everything seems fine now, including the push to the Docker Hub description https://hub.docker.com/r/freshrss/freshrss/

Ideas for future improvement:
* Look at https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml for inspiration, for instance regarding signing (let's first find out the pros and cons)
* Run https://github.com/FreshRSS/FreshRSS/actions/workflows/tests.yml first and kill the other actions if the tests fail
